### PR TITLE
Spark 3.5: Migrate remaining tests in source directory to JUnit5

### PR DESCRIPTION
**Pull request by chinmay-bhat**
This PR contains the remaining tests in spark/actions directory, now migrated to JUnit5.

Related to PR: #9342 

Issue: #9086

### CODE REVIEW

File: data/src/test/java/org/apache/iceberg/data/ParameterizedDeleteReadTest.java [hunk 1]
```diff
@@ -0,0 +1,591 @@
...
+import org.apache.iceberg.util.Pair;
+import org.apache.iceberg.util.StructLikeSet;
+import org.apache.iceberg.util.StructProjection;
+import org.junit.jupiter.api.AfterEach;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.TestTemplate;
+import org.junit.jupiter.api.io.TempDir;
+
+public abstract class ParameterizedDeleteReadTest {
+  // Schema passed to create tables
```
Comment (chinmay-bhat) [inline data/src/test/java/org/apache/iceberg/data/ParameterizedDeleteReadTest.java:51]: created a parameterized copy of DeleteReadTests since the classes that inherited from it were all parameterized and were in different modules (flink, mr, data).

**Comment by chinmay-bhat** [comment]
@nastra please review this PR too :) Thank you!

File: data/src/test/java/org/apache/iceberg/data/ParameterizedDeleteReadTest.java (hunk 1 above)
Comment (nastra) [reply data/src/test/java/org/apache/iceberg/data/ParameterizedDeleteReadTest.java:51]: in that case we should have a separate PR that moves `DeleteReadTests` and its subclasses to JUnit5 instead of creating a copy of the class

File: spark/v3.5/spark/src/test/java/org/apache/iceberg/spark/source/TestSparkScan.java [hunk 2]
```diff
@@ -57,39 +62,52 @@
...
+public class TestSparkScan extends TestBaseWithCatalog {
+
+  @Parameter(index = 3)
+  private String format;
+
+  @Parameters(name = "format = {0}")
+  public static Object[][] parameters() {
+    return new Object[][] {
+      {
+        SparkCatalogConfig.HADOOP.catalogName(),
```
Comment (nastra) [inline spark/v3.5/spark/src/test/java/org/apache/iceberg/spark/source/TestSparkScan.java:80]: is this correct? The old test parameterization was only using the file format

Comment (chinmay-bhat) [reply spark/v3.5/spark/src/test/java/org/apache/iceberg/spark/source/TestSparkScan.java:80]: Yup, this change is necessary since we removed constructors from the parameterized base classes. We now need to provide the values for `@Parameter` variables in the base class through the parameters() in the subclasses

File: data/src/test/java/org/apache/iceberg/data/ParameterizedDeleteReadTest.java (hunk 1 above)
Comment (chinmay-bhat) [reply data/src/test/java/org/apache/iceberg/data/ParameterizedDeleteReadTest.java:51]: ok, will create a new PR for this change

Comment (chinmay-bhat) [reply data/src/test/java/org/apache/iceberg/data/ParameterizedDeleteReadTest.java:51]: Hi @nastra, please see the [new PR here](https://github.com/apache/iceberg/pull/9382)

File: spark/v3.5/spark/src/test/java/org/apache/iceberg/spark/source/TestCompressionSettings.java [hunk 3]
```diff
@@ -69,66 +75,90 @@
...
 
   private static SparkSession spark = null;
 
-  private final FileFormat format;
-  private final ImmutableMap<String, String> properties;
+  @Parameter(index = 3)
+  private FileFormat format;
+
+  @Parameter(index = 4)
+  private ImmutableMap<String, String> properties;
```
Comment (nastra) [inline spark/v3.5/spark/src/test/java/org/apache/iceberg/spark/source/TestCompressionSettings.java:97]:
```suggestion
  private Map<String, String> properties;
```

File: spark/v3.5/spark/src/test/java/org/apache/iceberg/spark/source/TestFilteredScan.java [hunk 4]
```diff
@@ -286,7 +287,8 @@ public void testBucketPartitionedIDFilters() {
       InputPartition[] tasks = scan.planInputPartitions();
 
       // validate predicate push-down
-      Assert.assertEquals("Should create one task for a single bucket", 1, tasks.length);
+
```
Comment (nastra) [inline spark/v3.5/spark/src/test/java/org/apache/iceberg/spark/source/TestFilteredScan.java:290]: nit: unnecessary newline

File: spark/v3.5/spark/src/test/java/org/apache/iceberg/spark/source/TestFilteredScan.java [hunk 5]
```diff
@@ -314,7 +317,7 @@ public void testDayPartitionedTimestampFilters() {
       Batch scan = builder.build().toBatch();
 
       InputPartition[] tasks = scan.planInputPartitions();
-      Assert.assertEquals("Should create one task for 2017-12-21", 1, tasks.length);
+      assertThat(tasks).as("Should only create one task for 2017-12-21").hasSize(1);
```
Comment (nastra) [inline spark/v3.5/spark/src/test/java/org/apache/iceberg/spark/source/TestFilteredScan.java:320]:
```suggestion
      assertThat(tasks).as("Should create one task for 2017-12-21").hasSize(1);
```

File: spark/v3.5/spark/src/test/java/org/apache/iceberg/spark/source/TestPartitionValues.java [hunk 6]
```diff
@@ -102,35 +105,33 @@ public static Object[][] parameters() {
...
-  }
+  @Parameter(index = 1)
+  private boolean vectorized;
 
-  @Test
+  @TestTemplate
   public void testNullPartitionValue() throws Exception {
     String desc = "null_part";
-    File parent = temp.newFolder(desc);
+    File parent = new File(temp.toFile(), desc);
```
Comment (nastra) [inline spark/v3.5/spark/src/test/java/org/apache/iceberg/spark/source/TestPartitionValues.java:131]: should use same mechanism as other test classes in this PR to create a new folder

File: spark/v3.5/spark/src/test/java/org/apache/iceberg/spark/source/TestPartitionValues.java [hunk 7]
```diff
@@ -161,17 +162,17 @@ public void testNullPartitionValue() throws Exception {
...
+    assertThat(actual).as("Number of rows should match").hasSameSizeAs(expected);
+    assertThat(actual).as("Result rows should match").isEqualTo(expected);
   }
 
-  @Test
+  @TestTemplate
   public void testReorderedColumns() throws Exception {
     String desc = "reorder_columns";
-    File parent = temp.newFolder(desc);
+    File parent = new File(temp.toFile(), desc);
```
Comment (nastra) [inline spark/v3.5/spark/src/test/java/org/apache/iceberg/spark/source/TestPartitionValues.java:172]: same as above

File: spark/v3.5/spark/src/test/java/org/apache/iceberg/spark/source/TestPartitionValues.java (hunk 6 above)
Comment (nastra) [reply spark/v3.5/spark/src/test/java/org/apache/iceberg/spark/source/TestPartitionValues.java:131]: should this actually do `temp.resolve(desc)`?

File: spark/v3.5/spark/src/test/java/org/apache/iceberg/spark/source/TestPositionDeletesTable.java [hunk 8]
```diff
@@ -244,11 +244,14 @@ public void testSelect() throws IOException {
         };
     actual.sort(comp);
     expected.sort(comp);
-    assertEquals("Position Delete table should contain expected rows", expected, actual);
+    assertThat(actual)
+        .as("Position Delete table should contain expected rows")
+        .usingRecursiveComparison()
```
Comment (nastra) [inline spark/v3.5/spark/src/test/java/org/apache/iceberg/spark/source/TestPositionDeletesTable.java:249]: do we need this here?

File: spark/v3.5/spark/src/test/java/org/apache/iceberg/spark/source/TestPositionDeletesTable.java [hunk 9]
```diff
@@ -1051,24 +1086,24 @@ public void testWritePartitionEvolutionDisallowed() throws Exception {
...
-                .writeTo(posDeletesTableName)
-                .option(SparkWriteOptions.REWRITTEN_FILE_SCAN_TASK_SET_ID, fileSetID)
-                .append());
+    assertThatThrownBy(
+            () ->
+                scanDF
+                    .writeTo(posDeletesTableName)
+                    .option(SparkWriteOptions.REWRITTEN_FILE_SCAN_TASK_SET_ID, fileSetID)
+                    .append())
+        .isInstanceOf(AnalysisException.class);
```
Comment (nastra) [inline spark/v3.5/spark/src/test/java/org/apache/iceberg/spark/source/TestPositionDeletesTable.java:1101]: can you please add a `.hasMessage()` check here?

File: spark/v3.5/spark/src/test/java/org/apache/iceberg/spark/source/TestPositionDeletesTable.java [hunk 10]
```diff
@@ -1278,10 +1321,10 @@ public void testNormalWritesNotAllowed() throws IOException {
 
     Dataset<Row> scanDF = spark.read().format("iceberg").load(posDeletesTableName);
 
-    Assert.assertThrows(
-        "position_deletes table can only be written by RewriteDeleteFiles",
-        IllegalArgumentException.class,
-        () -> scanDF.writeTo(posDeletesTableName).append());
+    assertThatThrownBy(
+            () -> scanDF.writeTo(posDeletesTableName).append(),
+            "position_deletes table can only be written by RewriteDeleteFiles")
```
Comment (nastra) [inline spark/v3.5/spark/src/test/java/org/apache/iceberg/spark/source/TestPositionDeletesTable.java:1326]: I don't think we need this description. It should be self-explanatory that this isn't supported once we add a exception message check

File: spark/v3.5/spark/src/test/java/org/apache/iceberg/spark/source/TestReadProjection.java [hunk 11]
```diff
@@ -36,23 +38,19 @@
...
-import org.junit.Rule;
-import org.junit.Test;
-import org.junit.rules.TemporaryFolder;
+import org.junit.jupiter.api.TestTemplate;
+import org.junit.jupiter.api.extension.ExtendWith;
+import org.junit.jupiter.api.io.TempDir;
 
+@ExtendWith(ParameterizedTestExtension.class)
 public abstract class TestReadProjection {
-  final String format;
```
Comment (nastra) [inline spark/v3.5/spark/src/test/java/org/apache/iceberg/spark/source/TestReadProjection.java:44]: shouldn't this be marked with `@Parameter`?

File: spark/v3.5/spark/src/test/java/org/apache/iceberg/spark/source/TestSparkReadProjection.java [hunk 12]
```diff
@@ -51,40 +55,36 @@
...
+      {FileFormat.ORC, false, DISTRIBUTED},
+      {FileFormat.ORC, true, LOCAL}
     };
   }
 
-  private final FileFormat format;
-  private final boolean vectorized;
-  private final PlanningMode planningMode;
+  @Parameter(index = 0)
+  private FileFormat format;
```
Comment (nastra) [inline spark/v3.5/spark/src/test/java/org/apache/iceberg/spark/source/TestSparkReadProjection.java:79]: this should probably be defined in the base class rather than here

File: spark/v3.5/spark/src/test/java/org/apache/iceberg/spark/source/TestSparkReadProjection.java [hunk 13]
```diff
@@ -110,10 +110,10 @@ public static void stopSpark() {
   @Override
   protected Record writeAndRead(String desc, Schema writeSchema, Schema readSchema, Record record)
       throws IOException {
-    File parent = temp.newFolder(desc);
+    File parent = new File(temp.toFile(), desc);
```
Comment (nastra) [inline spark/v3.5/spark/src/test/java/org/apache/iceberg/spark/source/TestSparkReadProjection.java:113]: should this use .resolve like in the other classes?

File: spark/v3.5/spark/src/test/java/org/apache/iceberg/spark/source/TestSparkReaderDeletes.java [hunk 14]
```diff
@@ -81,32 +86,29 @@
...
-@RunWith(Parameterized.class)
-public class TestSparkReaderDeletes extends DeleteReadTests {
+import org.junit.jupiter.api.AfterAll;
+import org.junit.jupiter.api.AfterEach;
+import org.junit.jupiter.api.BeforeAll;
+import org.junit.jupiter.api.TestTemplate;
+import org.junit.jupiter.api.extension.ExtendWith;
+
+@ExtendWith(ParameterizedTestExtension.class)
+public class TestSparkReaderDeletes extends ParameterizedDeleteReadTest {
```
Comment (nastra) [inline spark/v3.5/spark/src/test/java/org/apache/iceberg/spark/source/TestSparkReaderDeletes.java:96]: `DeleteReadTests` and its subclasses should probably be handled in a separate PR

File: spark/v3.5/spark/src/test/java/org/apache/iceberg/spark/source/TestPartitionValues.java (hunk 6 above)
Comment (chinmay-bhat) [reply spark/v3.5/spark/src/test/java/org/apache/iceberg/spark/source/TestPartitionValues.java:131]: yup, it should have been `temp.resolve(desc)`.

When you read the below 3 lines together, at first glance it might seem like line 1 is doing something completely different compared to line 2 and 3, even though they're all creating `Files` hierarchically. To make it consistently readable, I chose `new File(temp.toFile(), desc)`

```
    File parent = temp.resolve(desc).toFile();
    File location = new File(parent, "test");
    File dataFolder = new File(location, "data");
```

File: spark/v3.5/spark/src/test/java/org/apache/iceberg/spark/source/TestReadProjection.java (hunk 11 above)
Comment (chinmay-bhat) [reply spark/v3.5/spark/src/test/java/org/apache/iceberg/spark/source/TestReadProjection.java:44]: `format` isn't actively being used anywhere in the file. It's only in the commented code of a few tests.
So I'm not sure if its needed

File: spark/v3.5/spark/src/test/java/org/apache/iceberg/spark/source/TestSparkReadProjection.java (hunk 12 above)
Comment (chinmay-bhat) [reply spark/v3.5/spark/src/test/java/org/apache/iceberg/spark/source/TestSparkReadProjection.java:79]: ok

File: spark/v3.5/spark/src/test/java/org/apache/iceberg/spark/source/TestPartitionValues.java (hunk 7 above)
Comment (chinmay-bhat) [reply spark/v3.5/spark/src/test/java/org/apache/iceberg/spark/source/TestPartitionValues.java:172]: https://github.com/apache/iceberg/pull/9380#discussion_r1446433055

File: spark/v3.5/spark/src/test/java/org/apache/iceberg/spark/source/TestSparkReadProjection.java (hunk 13 above)
Comment (chinmay-bhat) [reply spark/v3.5/spark/src/test/java/org/apache/iceberg/spark/source/TestSparkReadProjection.java:113]: https://github.com/apache/iceberg/pull/9380#discussion_r1446433055

File: spark/v3.5/spark/src/test/java/org/apache/iceberg/spark/source/TestSparkReaderDeletes.java (hunk 14 above)
Comment (chinmay-bhat) [reply spark/v3.5/spark/src/test/java/org/apache/iceberg/spark/source/TestSparkReaderDeletes.java:96]: As mentioned in https://github.com/apache/iceberg/pull/9380#discussion_r1436475714, I've opened a draft PR https://github.com/apache/iceberg/pull/9382 for exactly that. 

There's a blocking error in that PR I need help with. See https://github.com/apache/iceberg/pull/9382#issuecomment-1883659384.

File: spark/v3.5/spark/src/test/java/org/apache/iceberg/spark/source/TestSnapshotSelection.java [hunk 15]
```diff
@@ -44,57 +48,58 @@
...
-            TableProperties.DATA_PLANNING_MODE, planningMode.modeName(),
-            TableProperties.DELETE_PLANNING_MODE, planningMode.modeName());
-  }
+  @Parameter(index = 0)
+  private Map<String, String> properties;
 
-  @BeforeClass
-  public static void startSpark() {
+  @BeforeEach
+  public void startSpark() {
```
Comment (nastra) [inline spark/v3.5/spark/src/test/java/org/apache/iceberg/spark/source/TestSnapshotSelection.java:89]: previously this was started/stopped at the class-level. Any particular reason why we start/stop this now at the test method level?

File: spark/v3.5/spark/src/test/java/org/apache/iceberg/spark/source/TestPositionDeletesTable.java [hunk 16]
```diff
@@ -123,15 +126,9 @@ public static Object[][] parameters() {
...
   }
 
-  public TestPositionDeletesTable(
-      String catalogName, String implementation, Map<String, String> config, FileFormat format) {
-    super(catalogName, implementation, config);
-    this.format = format;
-  }
-
-  @Rule public TemporaryFolder temp = new TemporaryFolder();
+  @TempDir private Path temp;
```
Comment (nastra) [inline spark/v3.5/spark/src/test/java/org/apache/iceberg/spark/source/TestPositionDeletesTable.java:129]: is this needed? I think you should be able to use the one defined in the super class

File: spark/v3.5/spark/src/test/java/org/apache/iceberg/spark/source/TestSnapshotSelection.java (hunk 15 above)
Comment (chinmay-bhat) [reply spark/v3.5/spark/src/test/java/org/apache/iceberg/spark/source/TestSnapshotSelection.java:89]: no reason, changed it to `BeforeAll` and `AfterAll`

File: spark/v3.5/spark/src/test/java/org/apache/iceberg/spark/source/TestPositionDeletesTable.java (hunk 16 above)
Comment (chinmay-bhat) [reply spark/v3.5/spark/src/test/java/org/apache/iceberg/spark/source/TestPositionDeletesTable.java:129]: updated

**Review by nastra — APPROVED** [review]
LGTM, thanks @chinmay-bhat for getting this done!
